src/MSBuildBinLogQuery/Parse/Parser.cs

```diff
@@ -0,0 +1,208 @@
+                Consume<TaskToken>();
+
+                var next = ParseLogNodeOrNull();
+                return new TaskNode(next);
+            }
+
+            return ParseLogNodeWithType(LogNodeType.Direct);
+        }
+
+        private AstNode ParseTaskNodeOrLogNodeOrNull()
+        {
+            if (_scanner.Token is SingleSlashToken)
+            {
+                Consume<SingleSlashToken>();
+                return ParseTaskNodeOrSingleSlashLogNode();
+            }
+            else if (_scanner.Token is DoubleSlashToken)
+            {
+                Consume<DoubleSlashToken>();
+                return ParseLogNodeWithType(LogNodeType.All);
+            }
+
+            return null;
+        }
+
+        private AstNode ParseTargetNodeOrTaskNodeOrSingleSlashLogNode()
+        {
+            if (_scanner.Token is TargetToken)
+            {
+                Consume<TargetToken>();
+
+                var next = ParseTaskNodeOrLogNodeOrNull();
+                return new TargetNode(next);
+            }
+            else if (_scanner.Token is TaskToken)
+            {
+                Consume<TaskToken>();
+
+                var next = ParseLogNodeOrNull();
+                return new TaskNode(next);
+            }
+
+            return ParseLogNodeWithType(LogNodeType.Direct);
+        }
+
+        private AstNode ParseTargetNodeOrTaskNodeOrLogNodeOrNull()
+        {
+            if (_scanner.Token is SingleSlashToken)
+            {
+                Consume<SingleSlashToken>();
+                return ParseTargetNodeOrTaskNodeOrSingleSlashLogNode();
+            }
+            else if (_scanner.Token is DoubleSlashToken)
+            {
+                Consume<DoubleSlashToken>();
+                return ParseLogNodeWithType(LogNodeType.All);
+            }
+
+            return null;
+        }
+
+        private AstNode ParseProjectNodeOrTargetNodeOrTaskNodeOrSingleSlashLogNode()
+        {
+            if (_scanner.Token is ProjectToken)
+            {
+                Consume<ProjectToken>();
+
+                var next = ParseTargetNodeOrTaskNodeOrLogNodeOrNull();
+                return new ProjectNode(next is TaskNode ? next as TaskNode : next);
+            }
+            else if (_scanner.Token is TargetToken)
+            {
+                Consume<TargetToken>();
+
+                var next = ParseTaskNodeOrLogNodeOrNull();
+                return new TargetNode(next);
+            }
+            else if (_scanner.Token is TaskToken)
+            {
+                Consume<TaskToken>();
+
+                var next = ParseLogNodeOrNull();
+                return new TaskNode(next);
+            }
+
+            return ParseLogNodeWithType(LogNodeType.Direct);
+        }
+
+        private AstNode ParseQuery()
+        {
+            if (_scanner.Token is SingleSlashToken)
+            {
+                Consume<SingleSlashToken>();
+                return ParseProjectNodeOrTargetNodeOrTaskNodeOrSingleSlashLogNode();
```

Comment:
These names are a real mouthful. Is there one that's more meaningful in the domain of the MSBuild graph?

Reply:
Not sure for this one. But for others can we say `ParseNodeUnderProject` for `ParseTargetNodeOrTaskNodeOr...`

Reply:
Done.